src/plugins/axes/AxesConfig.cc

```diff
@@ -0,0 +1,379 @@
+}
+
+/////////////////////////////////////////////////
+void AxesConfig::UpdateActiveAxes()
+{
+  for (auto vis : this->dataPtr->activeAxesMap)
+  {
+    auto visAxes = std::dynamic_pointer_cast<rendering::AxisVisual>(
+          this->dataPtr->scene->VisualByName(vis.first + "Axes"));
+
+    auto visEntity = std::dynamic_pointer_cast<rendering::Visual>(
+          this->dataPtr->scene->VisualByName(vis.first));
+
+    if (visEntity && visAxes)
+    {
+      math::Pose3d pose_entity = visEntity->LocalPose();
+      visAxes->SetLocalPose(vis.second + pose_entity);
+    }
+  }
+}
+
+/////////////////////////////////////////////////
+void AxesConfig::UpdateOriginArrows()
+{
+  if (this->dataPtr->name_axes.empty())
+    return;
+
+  // Load axes if they don't already exist
+  this->LoadAxesbyName(this->dataPtr->name_axes);
+
+  // If axes were not loaded successfully, don't update
+  if (!this->dataPtr->axes)
+    return;
+
+  // Updating the poses
+  this->UpdateActiveAxes();
+
+  if (!this->dataPtr->dirty)
+    return;
+
+  // Save the axesVisual in the structure if it doesn't exist or
+  // update the pose
+  auto it = this->dataPtr->activeAxesMap.find(this->dataPtr->name_axes);
+  if (it == this->dataPtr->activeAxesMap.end())
+  {
+    this->dataPtr->activeAxesMap.insert(
+      std::pair<std::string, math::Pose3d>(
+        this->dataPtr->name_axes,
+        this->dataPtr->pose));
+  }
+  else
+  {
+    it->second = this->dataPtr->pose;
+  }
+
+  // update visibility
+  this->dataPtr->axes->SetVisible(this->dataPtr->visible);
+  // Update type: arrow or line
+  if (this->dataPtr->visible)
+  {
+    this->dataPtr->axes->ShowAxisHead(this->dataPtr->isArrow);
+  }
+  // update scale
+  this->dataPtr->axes->SetLocalScale(1, 1, this->dataPtr->length * 2);
+
+  this->dataPtr->dirty = false;
+}
+
+/////////////////////////////////////////////////
+const QStringList AxesConfig::comboList()
+{
+  return itemComboList;
+}
+
+/////////////////////////////////////////////////
+void AxesConfig::SetComboList(const QStringList &comboList)
+{
+  if (itemComboList != comboList)
+  {
+    itemComboList = comboList;
+    if (itemComboList.size() > 0 && this->dataPtr->name_axes.empty()) {
+      this->dataPtr->name_axes = itemComboList[0].toStdString();
+    }
+    emit ComboListChanged();
+  }
+}
+
+/////////////////////////////////////////////////
+void AxesConfig::LoadAxesbyName(const std::string & name)
+{
+  if ((this->dataPtr->name_axes.compare(name) == 0 && this->dataPtr->axes)
+      && !this->dataPtr->dirty)
+    return;
+
+  this->dataPtr->name_axes = name;
+  this->dataPtr->axes = std::dynamic_pointer_cast<rendering::AxisVisual>(
+    this->dataPtr->scene->VisualByName(this->dataPtr->name_axes + "Axes"));
+
+  if (!this->dataPtr->axes)
+  {
+    rendering::VisualPtr root = this->dataPtr->scene->RootVisual();
+
+    auto axes = this->dataPtr->scene->CreateAxisVisual(
+      this->dataPtr->name_axes + "Axes");
+    root->AddChild(axes);
```

Comment:
Why not add axes as a child of the visual it will be attached to? I think that should make it easier to update their pose.